mac/Keyman4MacIM/Keyman4MacIM/KMInputController.m

```diff
@@ -67,35 +58,74 @@ - (void)handleBackspace:(NSEvent *)event {
+  [self activateEventHandler:clientAppId];
 }
 
 /**
- * The user has switched to a different text input client: notification from KMInputMethodLifecycle
+ * Called by the app delegate when KMInputMethodLifecycle determines the user has deactivated Keyman
  */
-- (void)inputMethodChangedClient:(NSNotification *)notification {
-  os_log_debug([KMLogs lifecycleLog], "***KMInputController inputMethodChangedClient, deactivating old eventHandler and activating new one");
+- (void)deactivateClient {
+  os_log_debug([KMLogs lifecycleLog], "***KMInputController deactivateClient, deactivate old eventHandler");
+  
+  [self deactivateEventHandler];
+}
+
+- (void)deactivateEventHandler {
   if (_eventHandler != nil) {
+    os_log_debug([KMLogs lifecycleLog], " * KMInputController deactivate old eventHandler");
     [_eventHandler deactivate];
   }
-  _eventHandler = [[KMInputMethodEventHandler alloc] initWithClient:[KMInputMethodLifecycle getRunningApplicationId] client:self.client];
+}
 
+- (void)activateEventHandler:(NSString *)clientAppId {
+  os_log_debug([KMLogs lifecycleLog], " * KMInputController activate new eventHandler");
+  _eventHandler = [[KMInputMethodEventHandler alloc] initWithClient:clientAppId client:self.client];
 }
 
+/**
+ * Called by the OS to activate the input method, but some calls are not useful and are followed milliseconds later
+ * by a deactivateServer. These short-lived activations may be generated by clicking and releasing menu items without changing applications.
+ * Rather than treat every message received as a true activation, we send a message to KMInputMethodLifeCycle to evaluate.
+ */
 - (void)activateServer:(id)sender {
+  os_log_info([KMLogs lifecycleLog], " +++KMInputController, activateServer, self=%p", self);
   [sender overrideKeyboardWithKeyboardNamed:@"com.apple.keylayout.US"];
+  
+  /**
+   * When this KMInputController becomes the active server for the input method, then immediately update the AppDelegate.
+   * The duration that this controller is the active server may be extremely short, but, if so, we will receive another
+   * call to activateServer moments later and can update the AppDelegate again.
+   */
+  [self attachToAppDelegate];
+  
+  /**
+   * Call the shared lifecycle object so it can evaluate the current state and figure out whether this is a real activation of
+   * the input method or a change in clients or a false alarm.
+   */
   [KMInputMethodLifecycle.shared activateClient:sender];
 }
 
+- (void)attachToAppDelegate {
+  self.appDelegate.inputController = self;
+}
+
+/**
+ * Called by the OS to deactivate the input method
+ */
 - (void)deactivateServer:(id)sender {
+  os_log_info([KMLogs lifecycleLog], " +++KMInputController, deactivateServer, self=%p", self);
+
+  /**
+   * Call the shared lifecycle object so it can evaluate the current state and figure out whether this is a real deactivation of
+   * the input method or a change in clients or a false alarm.
+   */
```

Comment:
```suggestion
  /*
   Call the shared lifecycle object so it can evaluate the current state and figure out whether this is a real deactivation of
   the input method or a change in clients or a false alarm.
   */
```